File: app/src/main/java/kr/djspi/pipe01/fragment/ListDialog.java
package kr.djspi.pipe01.fragment;

import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.os.Parcelable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.fragment.app.DialogFragment;
import androidx.room.Room;

import com.sylversky.indexablelistview.scroller.Indexer;
import com.sylversky.indexablelistview.widget.IndexableListView;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import kr.djspi.pipe01.R;
import kr.djspi.pipe01.dto.PipeType.PipeTypeEnum;
import kr.djspi.pipe01.sql.Supervise;
import kr.djspi.pipe01.sql.SuperviseDatabase;

import static android.view.View.TEXT_ALIGNMENT_CENTER;
import static kr.djspi.pipe01.Const.PIPE_SHAPES;
import static kr.djspi.pipe01.Const.PIPE_TYPE_ENUMS;
import static kr.djspi.pipe01.Const.TAG_PIPE;
import static kr.djspi.pipe01.Const.TAG_SHAPE;
import static kr.djspi.pipe01.Const.TAG_SUPERVISE;

/**
 * 관로종류 목록과 관리기관 목록을 보여주는데 공용으로 사용하는 Dialog 클래스
 */
public class ListDialog extends DialogFragment implements OnClickListener {

    private static final String TAG = ListDialog.class.getSimpleName();
    private int selectIndex = -1;
    private String listTag;
    private String dialogTitle;
    private ArrayList<String> listItem;
    private IndexableListView listView;
    private String componentName;
    private Parcelable state;
    private OnSelectListener listener;
    private SuperviseDatabase superviseDb;

    public ListDialog() {
    }

    @Override
    public void onAttach(@NotNull Context context) {
        super.onAttach(context);
        listTag = getTag();
        listItem = new ArrayList<>();
        if (context instanceof OnSelectListener) {
            listener = (OnSelectListener) context;
        }
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        switch (listTag) {
            case TAG_PIPE:
                for (PipeTypeEnum pipe : PIPE_TYPE_ENUMS) {
                    listItem.add(pipe.getName());
                }
                dialogTitle = getString(R.string.popup_title_select_pipe);
                break;
            case TAG_SHAPE:
                listItem.addAll(Arrays.asList(PIPE_SHAPES));
                dialogTitle = getString(R.string.popup_title_select_shape);
                break;
            case TAG_SUPERVISE:
                new Thread(() -> {
                    if (superviseDb == null) {
                        superviseDb = Room.databaseBuilder(getContext(), SuperviseDatabase.class, "db_supervise").build();
                    }
                    List<Supervise> dbList = superviseDb.dao().getAll();
                    for (Supervise db : dbList) {
                        listItem.add(db.getSupervise());
                    }
                    superviseDb.close();
                }).start();
                dialogTitle = getString(R.string.popup_title_select_supervise);
                break;
            default:
                break;
        }
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_list, container, false);

        TextView titleView = view.findViewById(R.id.popup_title);
        titleView.setText(dialogTitle);
        view.findViewById(R.id.btn_ok).setOnClickListener(this);
        view.findViewById(R.id.btn_cancel).setOnClickListener(this);
        view.findViewById(R.id.btn_close).setOnClickListener(this);

        listView = view.findViewById(R.id.list_common);
//        listView.setIndexTypeface(Typeface.createFromAsset(context.getAssets(), "fonts/nanumsquareroundr.ttf"));
        if (listTag.equals(TAG_SUPERVISE)) {
            listView.setAdapter(new ListAdapter(getContext(), listItem, true));
        } else {
            listView.setAdapter(new ListAdapter(getContext(), listItem, false));
            listView.setFastScrollEnabled(false);
        }
        listView.setOnItemClickListener((parent, view1, position, id) -> {
            componentName = listItem.get(position);
            selectIndex = position;
        });
        if (state != null) {
            listView.requestFocus();
            listView.onRestoreInstanceState(state);
        }
        return view;
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.btn_ok:
                if (selectIndex == -1) {
                    Toast.makeText(getContext(), "항목을 선택해주세요", Toast.LENGTH_LONG).show();
                    return;
                }
                listener.onSelect(listTag, selectIndex, componentName);
                dismissAllowingStateLoss();
                break;
            case R.id.btn_cancel:
            case R.id.btn_close:
                dismissAllowingStateLoss();
            default:
                break;
        }
    }

    @Override
    public void onPause() {
        state = listView.onSaveInstanceState();
        super.onPause();
    }

    @Override
    public void onDismiss(@NotNull DialogInterface dialog) {
        selectIndex = -1;
        super.onDismiss(dialog);
    }

    /**
     * @see <a href="IndexableListView"></a>https://github.com/sylversky/IndexableListView.git</a>
     */
    private final class ListAdapter extends BaseAdapter implements Indexer {

        private final Context context;
        private final ArrayList<String> listItem;
        private final boolean isListSupervise;
        private final CustomSection customSection;

        ListAdapter(Context context, ArrayList<String> listItem, boolean isListSupervise) {
            this.context = context;
            this.listItem = listItem;
            this.isListSupervise = isListSupervise;
            this.customSection = isListSupervise ? new CustomSection(this) : null;
        }

        @Override
        public int getCount() {
            return listItem.size();
        }

        @Override
        public Object getItem(int position) {
            return listItem.get(position);
        }

        @Override
        public long getItemId(int position) {
            return position;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup container) {
            if (convertView == null) {
                convertView = LayoutInflater.from(context).inflate(R.layout.fragment_list_item, null);
            }
            TextView textView = convertView.findViewById(R.id.txt_name);
            textView.setText(listItem.get(position));
            textView.setTextAlignment(isListSupervise ? View.TEXT_ALIGNMENT_TEXT_START : TEXT_ALIGNMENT_CENTER);
            textView.setOnFocusChangeListener((v, hasFocus) -> {
            });
            return convertView;
        }

        @Override
        public String getComponentName(int position) {
            return listItem.get(position);
        }

        @Override
        @Nullable
        public Object[] getSections() {
            if (isListSupervise) return customSection.getArraySections();
            else return null;
        }

        @Override
        public int getPositionForSection(int sectionIndex) {
            return isListSupervise ? customSection.getPositionForSection(sectionIndex, getCount()) : 0;
        }

        @Override
        public int getSectionForPosition(int position) {
            return 0;
        }
    }
}

File: app/src/main/java/kr/djspi/pipe01/fragment/ListDialog.kt
package kr.djspi.pipe01.fragment

import android.content.Context
import android.content.DialogInterface
import android.graphics.Color
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnClickListener
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast
import androidx.fragment.app.DialogFragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import kr.djspi.pipe01.Const.TAG_SHAPE
import kr.djspi.pipe01.R
import kr.djspi.pipe01.databinding.FragmentListBinding
import kr.djspi.pipe01.dto.PipeShape

class ListDialog : DialogFragment(), OnClickListener {

    private var _binding: FragmentListBinding? = null
    private val binding get() = _binding!!

    private var listTag: String? = null
    private var dialogTitle: String? = null
    private var componentName: String? = null
    private var selectIndex: Int = -1
    private lateinit var listView: RecyclerView
    private lateinit var adapter: MyAdapter
    private lateinit var listItem: ArrayList<String>
    private lateinit var listener: OnSelectListener

    override fun onAttach(context: Context) {
        super.onAttach(context)
        listTag = tag
        listItem = ArrayList(0)
        if (context is OnSelectListener) {
            listener = context
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        when (listTag) {
//            TAG_PIPE -> {
//                Const.PIPE_TYPE_ENUMS.forEach {
//                    listItem.add(it.name)
//                }
//                dialogTitle = getString(R.string.popup_title_select_pipe)
//            }
            TAG_SHAPE -> {
                listItem.addAll(PipeShape.PipeShapeEnum.getSelectableTypes())
                dialogTitle = getString(R.string.popup_title_select_shape)
            }
//            TAG_SUPERVISE -> {
//                Thread(Runnable {
//                    BaseActivity.superviseDb!!.dao().all.forEach {
//                        listItem.add(it.supervise)
//                    }
//                }).start()
//                dialogTitle = getString(R.string.popup_title_select_supervise)
//            }
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentListBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val titleView = binding.popupListTitle
        titleView.text = dialogTitle
        listView = binding.listCommon
        listView.layoutManager = LinearLayoutManager(context)
        adapter = MyAdapter(listItem) { selectedItem ->
            this.componentName = selectedItem
            this.selectIndex = listItem.indexOf(componentName)
        }
        binding.btnOk.setOnClickListener(this)
        binding.btnCancel.setOnClickListener(this)
        binding.buttonClose.setOnClickListener(this)
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.btn_ok -> {
                if (selectIndex == -1) {
                    Toast.makeText(context, "항목을 선택해주세요", Toast.LENGTH_LONG).show()
                    return
                }
                listener.onSelect(listTag, selectIndex, componentName)
                dismissAllowingStateLoss()
            }
            R.id.btn_cancel, R.id.button_close -> dismissAllowingStateLoss()
        }
    }

    override fun onPause() {
        super.onPause()
        listView.adapter = null
    }

    override fun onResume() {
        super.onResume()
        listView.adapter = adapter
    }

    override fun onDismiss(dialog: DialogInterface) {
        selectIndex = -1
        super.onDismiss(dialog)
    }
}

class MyAdapter(
    private val dataSet: ArrayList<String>,
    private val onItemSelected: (String) -> Unit,
): RecyclerView.Adapter<MyAdapter.ViewHolder>() {

    private var selectedItemPosition = -1
    private var selectedLayout: LinearLayout? = null

    inner class ViewHolder(view: View): RecyclerView.ViewHolder(view) {
        val itemLayout: LinearLayout = view.findViewById(R.id.item_root)
        val itemText: TextView = view.findViewById(R.id.item_text)

        fun bind(item: String) {
            itemText.text = item
        }
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(viewGroup.context).inflate(R.layout.fragment_list_item, viewGroup, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(viewHolder: ViewHolder, position: Int) {
        val item = dataSet[position]
        val layout = viewHolder.itemLayout
        viewHolder.bind(item)

        layout.setOnClickListener {
            val currentPosition = viewHolder.adapterPosition
            if (selectedItemPosition >= 0 || selectedLayout != null) {
                selectedLayout?.setBackgroundColor(Color.TRANSPARENT)
            }

            selectedItemPosition = currentPosition
            selectedLayout = layout
            selectedLayout?.setBackgroundColor(Color.parseColor("#ffccbc"))
            onItemSelected(item)
        }
    }

    override fun getItemCount(): Int = dataSet.size
}
